CONTRIBUTING.md

```diff
@@ -0,0 +1,295 @@
+- Write/update documentation when necessary.
+- When creating a test that is linked to a GitHub issue, add the number at the back with a comma, as in the following:
+
+```fsharp
+[<Test>]
+let ``preserve compile directive between piped functions, 512`` () = ...
+```
+
+### Small steps
+
+It is better to create a draft pull request with some initial small changes, and engage conversation, than to spend a lot of effort on a large pull request that was never discussed.
+Someone might be able to warn you in advance that your change will have wide implications for the rest of Fantomas, or might be able to point you in the right direction;
+but this can only happen if you discuss your proposed changes early and often.
+It's often better to check *before* contributing that you're setting off on the right path.
+
+# Your First Contribution
+
+> “It's a dangerous business, Frodo, going out of your door," he used to say. "You step into the Road, and if you don't keep your feet, there is no knowing where you might be swept off to. ― J.R.R. Tolkien, The Fellowship of the Ring
+
+Fantomas is a project that has its roots deeply nested in the F# compiler. This can be an overwhelming experience at first, and it might even make you nervous about contributing in the first place.
+Fear not: once you get the hang of it, things are less complicated than they seem.
+
+In short, Fantomas is a source-code-to-source-code compiler. It will transform the text in the source code to an intermediate format and transform that again to source code.
+It uses the [F# Compiler Services](https://fsharp.github.io/FSharp.Compiler.Service/index.html) to do this. The F# compiler will be used to create an [UnTyped Abstract Syntax](https://fsharp.github.io/FSharp.Compiler.Service/reference/fsharp-compiler-syntaxtree.html) tree (or "AST").
+The AST is then reprinted in `CodePrinter.fs`: once the whole tree is traversed, the formatted code can be constructed.
+
+There is a catch, though. The AST does not have all the original information the was once in the source code. 
+Things like code comments, for example, are not present in the tree given to us by the F# compiler.
+We call these pieces of information "trivia".
+Fantomas uses another part of the F# Compiler Services to detect trivia: namely, the tokenizer.
+All tokens are processed to find trivia, then trivia are assigned to "trivia nodes" (a concept of Fantomas, not of the F# compiler).
+Trivia nodes are linked to an existing AST node or a F# token. They serve as a marker to indicate where in `CodePrinter.fs` the trivia needs to be added to the formatted code.
+
+## YouTube video series
+
+In order to better understand how everything works, you can watch a [YouTube video series](https://www.youtube.com/playlist?list=PLvw_J2kfZCX3Mf6tEbIPZXbzJOD1VGl4K) on how Fantomas internally works.
+This is highly recommended for anybody new to the code base.
```

Comment:
Is code base two words or one?  I'm never quite sure, but I think I see it mostly as one.